File: frontend/src/services/release.ts
import axios from "axios";

import { api } from "./api";

import type {
  GenerateReleaseNotesInput,
  ReleaseNotes,
} from "../types/release";


type ReleaseNotesApiResponse = {
  title: string;
  summary: string;

  features: string[];
  fixes: string[];
  improvements: string[];
  documentation: string[];
  maintenance: string[];

  contributors: string[];

  total_commits: number;
};


function getErrorMessage(
  error: unknown,
): string {
  if (axios.isAxiosError(error)) {
    const detail = error.response?.data?.detail;

    if (
      typeof detail === "string" &&
      detail.trim()
    ) {
      return detail;
    }

    if (error.code === "ERR_NETWORK") {
      return (
        "Unable to reach the release generation API."
      );
    }
  }

  if (error instanceof Error) {
    return error.message;
  }

  return "Unable to generate release notes.";
}


export async function generateReleaseNotes(
  input: GenerateReleaseNotesInput,
): Promise<ReleaseNotes> {
  try {
    const response =
      await api.post<ReleaseNotesApiResponse>(
        "/releases/generate",
        {
          repository: input.repository,
          branch: input.branch,

          title: input.title,
          version: input.version,
          environment: input.environment,

          generate_for: input.generateFor,

          selected_author:
            input.selectedAuthor,

          commits: input.commits.map(
            (commit) => ({
              id: commit.id,
              message: commit.message,
              author: commit.author,

              author_email:
                commit.authorEmail,

              date: commit.date,
            }),
          ),
        },
      );

    return {
      title: response.data.title,
      summary: response.data.summary,

      features: response.data.features,
      fixes: response.data.fixes,

      improvements:
        response.data.improvements,

      documentation:
        response.data.documentation,

      maintenance:
        response.data.maintenance,

      contributors:
        response.data.contributors,

      totalCommits:
        response.data.total_commits,
    };
  } catch (error) {
    throw new Error(
      getErrorMessage(error),
    );
  }
}

File: backend/app/schemas/release.py
from typing import Literal

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
)


ReleaseTarget = Literal[
    "all",
    "individual",
]


class ReleaseCommit(BaseModel):
    id: str = Field(
        ...,
        min_length=1,
    )

    message: str = Field(
        ...,
        min_length=1,
    )

    author: str = Field(
        ...,
        min_length=1,
    )

    author_email: str = ""
    date: str = ""


class ReleaseGenerateRequest(BaseModel):
    repository: str = Field(
        ...,
        min_length=1,
    )

    branch: str = Field(
        ...,
        min_length=1,
    )

    title: str = Field(
        ...,
        min_length=1,
    )

    version: str = ""

    environment: str = Field(
        ...,
        min_length=1,
    )

    generate_for: ReleaseTarget = "all"

    selected_author: str = ""

    commits: list[ReleaseCommit] = Field(
        ...,
        min_length=1,
        max_length=100,
    )


class AIReleaseItem(BaseModel):
    model_config = ConfigDict(
        extra="forbid",
    )

    text: str = Field(
        ...,
        min_length=1,
        max_length=500,
    )

    commit_ids: list[str] = Field(
        ...,
        min_length=1,
        max_length=100,
    )


class AIReleaseContent(BaseModel):
    model_config = ConfigDict(
        extra="forbid",
    )

    summary: str = Field(
        ...,
        min_length=1,
        max_length=1500,
    )

    features: list[AIReleaseItem]
    fixes: list[AIReleaseItem]
    improvements: list[AIReleaseItem]
    documentation: list[AIReleaseItem]
    maintenance: list[AIReleaseItem]


class ReleaseNotesResponse(BaseModel):
    title: str
    summary: str

    features: list[str]
    fixes: list[str]
    improvements: list[str]
    documentation: list[str]
    maintenance: list[str]

    contributors: list[str]

    total_commits: int
